tornado/locks.py

```diff
@@ -241,3 +241,63 @@ def release(self):
         if self._value >= self._initial_value:
             raise ValueError("Semaphore released too many times")
         super(BoundedSemaphore, self).release()
+
+
+class Lock(object):
+    """A lock for coroutines.
+
+    A Lock begins unlocked, and `acquire` locks it immediately. While it is
+    locked, a coroutine that yields `acquire` waits until another coroutine
+    calls `release`.
+
+    Releasing an unlocked lock raises `RuntimeError`.
+
+    `acquire` supports the context manager protocol:
+
+    >>> from tornado import gen, locks
+    >>> lock = locks.Lock()
+    >>>
+    >>> @gen.coroutine
+    ... def f():
+    ...    with (yield lock.acquire()):
+    ...        # Do something holding the lock.
+    ...        pass
+    ...
+    ...    # Now the lock is released.
+
+    Coroutines waiting for `acquire` are granted the lock in first-in, first-out
```

Comment:
OK, I put it back, this time with list(range(N)) to fix the Python 3 test
bug.


On Sat, Feb 28, 2015 at 1:36 PM, Ben Darnell <notifications@github.com>
wrote:

> In tornado/locks.py
> <https://github.com/tornadoweb/tornado/pull/1353#discussion_r25559864>:
>
> > +    Releasing an unlocked lock raises `RuntimeError`.
> > +
> > +    `acquire` supports the context manager protocol:
> > +
> > +    >>> from tornado import gen, locks
> > +    >>> lock = locks.Lock()
> > +    >>>
> > +    >>> @gen.coroutine
> > +    ... def f():
> > +    ...    with (yield lock.acquire()):
> > +    ...        # Do something holding the lock.
> > +    ...        pass
> > +    ...
> > +    ...    # Now the lock is released.
> > +
> > +    Coroutines waiting for `acquire` are granted the lock in first-in, first-out
>
> I'm OK with keeping the test, actually - we can remove it when and if we
> decide to change the implementation, and in the meantime it will make sure
> we don't accidentally break things.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tornadoweb/tornado/pull/1353/files#r25559864>.
>